src/Tools/dotnet-monitor/StartupHook/StartupHookService.cs

```diff
@@ -0,0 +1,59 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Microsoft.Diagnostics.Monitoring.WebApi;
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Diagnostics.Tools.Monitor.StartupHook
+{
+    internal sealed class StartupHookService :
+        IDiagnosticLifetimeService
+    {
+        private readonly IEndpointInfo _endpointInfo;
+        private readonly IInProcessFeatures _inProcessFeatures;
+        private readonly TaskCompletionSource<bool> _resultSource = new(TaskCreationOptions.RunContinuationsAsynchronously);
+        private readonly StartupHookValidator _startupHookValidator;
+
+        public StartupHookService(
+            IEndpointInfo endpointInfo,
+            IInProcessFeatures inProcessFeatures,
+            StartupHookValidator startupHookValidator)
+        {
+            _endpointInfo = endpointInfo ?? throw new ArgumentNullException(nameof(endpointInfo));
+            _inProcessFeatures = inProcessFeatures ?? throw new ArgumentNullException(nameof(inProcessFeatures));
+            _startupHookValidator = startupHookValidator ?? throw new ArgumentNullException(nameof(startupHookValidator));
+        }
+
+        public async ValueTask StartAsync(CancellationToken cancellationToken)
```

Comment:
Just as a sanity check since I had to modify this logic the last time - can this only be called once? Since we're not verifying that we haven't already tried to apply the startup hook.

Reply:
Yes, it's only called once for the lifetime of the diagnostic session. If the target app fails diagnostic commands and continuously disconnects and reconnects for some reason, then it will show up each time.

I have an idea of storing some data about each process after each session is ended, so this could be added to that in order to prevent continuous log writing if need be.